### PR TITLE
feat(governance): add MandateGate context guard (#1868)

### PR DESCRIPTION
**Pull request by fahertym**
## Update (post-review, 2026-05-28)

Codex P1 on `server.rs` was correct: extending
`missing_production_dependencies()` to include `mandate_gate` made
`GovernanceContext::validate()` reject Production startup
unconditionally on the gateway path, because the daemon-backed
construction left the field `None`. Fixed in 5a02775.

The gateway's daemon-backed path now wires a real `DefaultMandateGate`
from the existing `receipt_store` (which already implements
`GovernanceReceiptBackend`). No new persistence; no stub gates in
production code. Bootstrap/Test construction sites elsewhere remain
`None` to exercise the warn-only fall-open posture.

Regression test added at
`icn/crates/icn-gateway/tests/mandate_gate_production_wiring.rs`:

- `production_gateway_context_validates_when_mandate_gate_wired_from_receipt_store`
  — gateway-shaped Production context with a real `DefaultMandateGate`
  over `ReceiptStore` passes `validate()`.
- `production_gateway_context_still_rejects_when_mandate_gate_missing`
  — same shape with `mandate_gate: None` still fails closed, naming
  `mandate_gate` in `err.missing`. Both assertions live in one file so
  any future weakening of the guard shows up as a test diff here.

Also addressed a Copilot review nit on the build_mode test's negative
assertion (a31c7f3) — replaced an ends_with check that could never
fire with a tighter contains check on the snake_case identifier.

Non-claims preserved:

- No handler calls `MandateGate::require()` yet (#1868 step 7 is still
  out of scope).
- No receipt-body schema migration.
- No `governance:write` retirement.
- No kernel meaning-firewall widening — the gateway construction site
  spells the trait, never the act-time semantics.
- No production-readiness, live-federation, or demo-ready claim. The
  Production startup guard is enabled and satisfiable on the gateway
  path; nothing here makes the broader system ready to deploy.

---

## What

Adds `mandate_gate: Option<Arc<dyn MandateGate + Send + Sync>>` to
`GovernanceContext<E>` in `icn/apps/governance/src/http/configure.rs`
and integrates it into the existing
`missing_production_dependencies()` / `validate()` machinery, mirroring
the established pattern for `member_checker`, `suspension_checker`,
and `membership_resolver`.

Behavior matrix:

| Build mode  | `mandate_gate: None`         | `mandate_gate: Some(_)` |
|-------------|------------------------------|-------------------------|
| Production  | `validate()` **fails closed**, naming `mandate_gate` in the error | `validate()` returns `Ok(())` |
| Bootstrap   | warning, `validate()` returns `Ok(())` | `validate()` returns `Ok(())` |
| Test        | warning, `validate()` returns `Ok(())` | `validate()` returns `Ok(())` |

Construction-site changes are mechanical: 48 `GovernanceContext { … }`
literals across the workspace receive `mandate_gate: None,`. No
handler behavior change.

## Why this is the next #1868 slice

PR #1926 merged the step-6 `MandateGate` resolver primitives as a
purely additive module — its own docs state "Nothing calls this gate
in step 6." `docs/design/governance/mandate-gate-design.md` §9 names
two follow-up directions:

1. **Step 7 — handler wiring**: wire `require()` into the charter and
   federation handlers, recording a `MandateGrant` hash in receipts
   (depends on the step-2 receipt schema).
2. **Startup guard (with #1871)**: extend `GovernanceContext::validate()`
   so Production refuses to start without a wired gate.

This PR is the **Startup guard** slice. It is intentionally narrower
than step 7 so the act-time mandate semantics can be reviewed
independently of the infrastructure plumbing.

There are 49 `GovernanceContext` construction sites in the workspace.
Combining the field addition with handler behavior change for
`activate_charter` would push the diff past the small/reviewable
boundary; the design doc also recommends wiring charter and federation
together in step 7, not `activate_charter` alone.

## Tests

Added in `build_mode_tests` in `configure.rs`:

- `production_without_mandate_gate_rejects` — all three
  standing/resolver deps wired, `mandate_gate: None` → `validate()`
  returns `Err(...)` and `missing` contains exactly `"mandate_gate"`;
  the rendered message references MandateGate and the doctrine
  ("capability scope is not a mandate").
- `bootstrap_without_mandate_gate_does_not_reject`,
  `test_mode_without_mandate_gate_does_not_reject` — Bootstrap/Test
  stay permissive.
- Existing tests
  (`missing_production_dependencies_lists_unset_optional_fields`,
   `missing_production_dependencies_excludes_wired_fields`,
   `missing_production_dependencies_empty_when_all_wired`,
   `production_all_missing_returns_error_naming_all`,
   `production_single_missing_dep_names_only_that_dep`,
   `production_all_wired_validates_ok`) extended to include
  `mandate_gate` in their assertions, preserving the stable-ordering
  contract on the missing-deps list.

A test-only `DummyMandateGate` lives inside `#[cfg(test)]` and always
returns `Err(Rejected(NoMandate))`. It is not exported; nothing in
production code can construct it.

## Invariants preserved

- **Meaning firewall.** `MandateGate`, `MandateAct`, `MandateTarget`
  stay in `apps/governance`. The kernel sees nothing new. The field
  stored in `GovernanceContext` is an app-side handle.
- **Composability.** `mandate_gate` is a sibling of `member_checker` /
  `suspension_checker` / `membership_resolver`, not a wrapper. No
  god-auth function. The async `suspension_checker` stays
  handler-composed (the gate is synchronous by design).
- **Trait surface untouched.** Using `Arc<dyn MandateGate + Send + Sync>`
  at the field type adds the marker bounds at the coercion point only.
  The `MandateGate` trait keeps its current definition.
- **Behavior-neutral for Bootstrap/Test.** Every existing construction
  site gets `mandate_gate: None`. `validate()` in Bootstrap/Test
  continues to return `Ok(())`. No handler calls the gate.

## Validation run

All commands from `icn/icn/` unless noted:

- `cargo fmt --all -- --check` — pass
- `cargo check --workspace --all-targets` — pass
- `cargo clippy -p icn-governance-actor --all-targets -- -D warnings` — pass
- `cargo test -p icn-governance-actor` — pass (270 lib unit, plus all integration suites)
- `cargo test -p icn-gateway --tests` — pass (35 test groups, all green)
- `python3 .github/scripts/compliance_linter.py` (from repo root) — pass (107 files scanned, no violations)
- `git diff --check` — clean

## Out of scope / non-claims

- Does **not** wire any handler. `activate_charter` and all other
  handlers are untouched. Step 7 will wire charter and federation
  handlers together in a follow-up PR.
- Does **not** change receipt-body schema (no `MandateGrantRef`, no
  `capability_scope_presented`, no receipt v2).
- Does **not** retire `governance:write`.
- Does **not** move mandate semantics into kernel crates. The meaning
  firewall is unchanged.
- Does **not** change public surfaces, demos, the website, K3s
  manifests, or partner/NYCN activation paths.
- Does **not** claim production readiness, live-federation readiness,
  or any deployment milestone. This slice only proves the startup
  guard.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: icn/crates/icn-gateway/src/server.rs [hunk 1]
```diff
@@ -1416,6 +1416,7 @@ impl GatewayServer {
             // Daemon mode: SDIS execution goes through the actor event system
             // (KernelGovernanceExecutor → SdisServiceImpl). HTTP path leaves this None.
             sdis_service: None,
+            mandate_gate: None,
```
[the remark this answers is not in the document]
Comment (fahertym) [reply icn/crates/icn-gateway/src/server.rs]: P1 was correct — thanks. Adding `mandate_gate` to `missing_production_dependencies()` turned the `Production` startup guard on without giving the gateway path a way to satisfy it, so `validate()` was rejecting unconditionally on `mandate_gate` rather than enforcing a real check.

Fixed in 5a02775. The daemon-backed gateway path now wires a real `DefaultMandateGate` over the existing `receipt_store` (which already implements `GovernanceReceiptBackend`) — no new persistence, no stub gate in production code. Production still fails closed if `mandate_gate` is missing; a regression test at `icn/crates/icn-gateway/tests/mandate_gate_production_wiring.rs` keeps both directions honest:

- `production_gateway_context_validates_when_mandate_gate_wired_from_receipt_store` — gateway-shaped Production context with a real `DefaultMandateGate` validates.
- `production_gateway_context_still_rejects_when_mandate_gate_missing` — same shape with `mandate_gate: None` still rejects, naming `mandate_gate` in `err.missing`.

No handler calls `MandateGate::require()` yet — handler wiring (#1868 step 7) remains out of scope for this PR. Bootstrap/Test sites elsewhere intentionally stay `None` to exercise the warn-only posture.

File: icn/apps/governance/src/http/configure.rs [hunk 2]
```diff
@@ -953,6 +1017,10 @@ mod build_mode_tests {
             !msg.contains("member_checker,"),
             "must not falsely name wired dep: {msg}"
         );
+        assert!(
+            !msg.contains("mandate_gate,") && !msg.ends_with("mandate_gate."),
```
[the remark this answers is not in the document]
Comment (fahertym) [reply icn/apps/governance/src/http/configure.rs]: Good catch — addressed in a31c7f3. Replaced the broken `!msg.ends_with("mandate_gate.")` half with `!msg.contains("mandate_gate")`. The snake_case identifier can only enter the message via `missing.join(", ")`, since the doctrine line uses CamelCase `MandateGate`; that makes the contains check both sound (works in any list position) and tight (no false positive on the doctrine text). The strong assertion remains `err.missing == ["membership_resolver"]` directly above this — the string check is just a sanity guard.
